base/sort.jl

```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+
+        v, t = t, v # swap the now sorted destination list t back into the primary list v
+
+    end
+
+    v
+end
+
+function sort!(v::AbstractVector{<:Bool}, lo::Integer, hi::Integer, a::AdaptiveSort, o::Ordering)
+    first = lt(o, false, true) ? false : lt(o, true, false) ? true : return v
+    count = 0
+    for i in lo:hi
+        if v[i] == first
+            count += 1
+        end
+    end
+    v[lo:lo+count-1] .= first
+    v[lo+count:hi] .= !first
+    v
+end
+
+maybe_unsigned(x::Integer) = x # this is necessary to avoid calling unsigned on BigInt
+maybe_unsigned(x::Union{Int8, Int16, Int32, Int64, Int128}) = unsigned(x)
+function _extrema(v::AbstractArray, lo::Integer, hi::Integer, o::Ordering)
+    mn = mx = v[lo]
+    while lo < hi
+        lo += 1
+        vi = v[lo]
+        lt(o, vi, mn) && (mn = vi)
+        lt(o, mx, vi) && (mx = vi)
+    end
+    mn, mx
+end
+function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::AdaptiveSort, o::Ordering)
+    # if the sorting task is unserializable, then we can't radix sort or sort_int_range!
+    # so we skip straight to the fallback algorithm which is comparison based.
+    U = Serial.Serializable(eltype(v), o)
+    U === nothing && return sort!(v, lo, hi, a.fallback, o)
+
+    # to avoid introducing excessive detection costs for the trivial sorting problem
+    # and to avoid overflow, we check for small inputs before any other runtime checks
+    hi <= lo && return v
+    ln = maybe_unsigned(hi-lo) # ln == length(v[lo:hi])-1. Lower number = fewer letters :)
+    # only count sort on a short range can compete with insertion sort fo ln < 30
+    # and the optimization is not worth the detection cost, so we use insertion sort.
+    ln < 30 && return sort!(v, lo, hi, SMALL_ALGORITHM, o)
+
+    # UInt128 does not support fast bit shifting so we never
+    # dispatch to radix sort but we may still perform count sort
+    if sizeof(U) > 8
+        if eltype(v) <: Integer && o isa DirectOrdering
+            mn, mx = _extrema(v, lo, hi, Forward)
+            rln = maybe_unsigned(mx-mn)
+
+            # we know ln ≥ 30, so this will never underflow.
+            # if ln > 3.7e18 (59 exabytes), then this may incorrectly dispatch to fallback
+            if rln < 5ln-100 # count sort will outperform comparison sort if rln is small
+                return sort_int_range!(v, rln+1, mn, o === Forward ? identity : reverse, lo, hi)
+            end
+        end
+        return sort!(v, lo, hi, a.fallback, o)
+    end
+
+    mn, mx = _extrema(v, lo, hi, o)
+    if eltype(v) <: Integer && o isa DirectOrdering
+        R = o === Reverse
+        rln = maybe_unsigned(R ? mn-mx : mx-mn)
+        if rln < div(ln, 2) # count sort will be superior if rln is very small
+            return sort_int_range!(v, rln+1, R ? mx : mn, R ? reverse : identity, lo, hi)
+        end
+    end
+
+    umn, umx = Serial.serialize(mn, o), Serial.serialize(mx, o)
```

Comment:
why are we serializing?

Reply:
To compute `umx-umn` which informs the decision about whether to use `InsertionSort` or `sort_int_range!` or `RadixSort`. If `umx-umn` is small, then value based methods are more likely to outperform comparison based methods.

Reply:
The variables `umn` and `umx` are the extrema of the array `u`, but the decision about whether to compute `u` at all depends on `umn` and `umx`. It's backward from a readability standpoint.

Reply:
makes sense.